File: basic_dsa_three_hundred_two.cpp
/**
 * Basic DSA - Missing Number (LeetCode 268)
 * Find the missing number in range [0..n] from array of length n.
 */
#include <iostream>
#include <vector>

int missingNumber(std::vector<int>& nums) {
    int n = nums.size();
    int x = n;
    for (int i = 0; i < n; i++) x ^= i ^ nums[i];
    return x;
}

int main() {
    std::cout << std::string(60, '=') << "\n";
    std::cout << "Problem: Missing Number\n";
    std::cout << std::string(60, '=') << "\n";

    std::vector<int> v1 = {3, 0, 1};
    std::cout << "[3,0,1] -> " << missingNumber(v1) << "\n";

    std::vector<int> v2 = {0, 1};
    std::cout << "[0,1] -> " << missingNumber(v2) << "\n";

    std::vector<int> v3 = {9, 6, 4, 2, 3, 5, 7, 0, 1};
    std::cout << "n=9 array -> " << missingNumber(v3) << "\n";
}

File: basic_dsa_three_hundred_two.py
"""
Basic DSA - Range Sum Query - Immutable
Given an integer array nums, handle multiple queries of the following type:
Return the sum of the elements of nums between indices left and right inclusive.
"""


class NumArray:
    """
    Prefix sums.
    Build: O(n)
    Query: O(1)
    Space: O(n)
    """

    def __init__(self, nums):
        self.pref = [0]
        running = 0
        for x in nums:
            running += x
            self.pref.append(running)

    def sum_range(self, left, right):
        return self.pref[right + 1] - self.pref[left]


if __name__ == "__main__":
    print("=" * 60)
    print("Problem: Range Sum Query - Immutable")
    print("=" * 60)

    num_array = NumArray([-2, 0, 3, -5, 2, -1])
    print(num_array.sum_range(0, 2))  # 1
    print(num_array.sum_range(2, 5))  # -1
    print(num_array.sum_range(0, 5))  # -3
